packages/nouns-contracts/contracts/NounsAuctionHouseV2.sol

```diff
@@ -0,0 +1,511 @@
+            }
+        }
+    }
+
+    /**
+     * @notice Get past auction settlements.
+     * @dev Returns settlements in reverse order, meaning settlements[0] will be the most recent auction price.
+     * Skips auctions where there was no winner, i.e. no bids.
+     * @param auctionCount The number of price observations to get.
+     * @return settlements An array of type `Settlement`, where each Settlement includes a timestamp,
+     * the Noun ID of that auction, the winning bid amount, and the winner's address.
+     */
+    function getSettlements(uint256 auctionCount) external view returns (Settlement[] memory settlements) {
+        uint256 latestNounId = auctionStorage.nounId;
+        if (!auctionStorage.settled && latestNounId > 0) {
+            latestNounId -= 1;
+        }
+
+        settlements = new Settlement[](auctionCount);
+        uint256 actualCount = 0;
+        while (actualCount < auctionCount && latestNounId > 0) {
+            SettlementState memory settlementState = settlementHistory[latestNounId];
+            // Skip Nouner reward Nouns, they have no price
+            // Also skips IDs with no price data
+            if (settlementState.winner == address(0)) {
+                --latestNounId;
+                continue;
+            }
+
+            settlements[actualCount] = Settlement({
+                blockTimestamp: settlementState.blockTimestamp,
+                amount: uint64PriceToUint256(settlementState.amount),
+                winner: settlementState.winner,
+                nounId: latestNounId
+            });
+            ++actualCount;
+            --latestNounId;
+        }
+
+        if (auctionCount > actualCount) {
+            // this assembly trims the observations array, getting rid of unused cells
+            assembly {
+                mstore(settlements, actualCount)
+            }
+        }
+    }
+
+    /**
+     * @notice Get past auction prices.
+     * @dev Returns prices in reverse order, meaning prices[0] will be the most recent auction price.
+     * Skips auctions where there was no winner, i.e. no bids.
+     * @param auctionCount The number of price observations to get.
+     * @return prices An array of uint256 prices.
+     */
+    function getPrices(uint256 auctionCount) external view returns (uint256[] memory prices) {
+        uint256 latestNounId = auctionStorage.nounId;
+        if (!auctionStorage.settled && latestNounId > 0) {
+            latestNounId -= 1;
+        }
+
+        prices = new uint256[](auctionCount);
+        uint256 actualCount = 0;
+        while (actualCount < auctionCount && latestNounId > 0) {
+            SettlementState memory settlementState = settlementHistory[latestNounId];
+            // Skip Nouner reward Nouns, they have no price
+            // Also skips IDs with no price data
+            if (settlementState.winner == address(0)) {
+                --latestNounId;
+                continue;
+            }
```

Comment:
I'm not so sure that we should skip auctioned nouns with no price data. Any gaps in historical population may return unexpected price results that could in turn affect the burn.

Reply:
let's discuss this

Reply:
@solimander At first I thought you meant nouns that are auctioned but got no bids, and therefore burned.
But now I think I understand you meant historical data gaps for which we didn't populate data.

Trying to build a more concrete example:
* We deploy this and start saving data starting from noun 900
* We populate past data for nouns 800-850
* Latest noun is at 925 and another contract calls `getPrices(50)`, then they get the prices for nouns 825-850 & 900-925 (ignoring nounder rewards for now), which is not what they're expecting.

1. Is this what you meant? if not, could you provide an example that demonstrates the issue?
2. Thinking what would be a better behavior in this case. Some options:
  a. Don't skip the ids, but also don't return data for them. e.g. for `getPrices(50)`, returns only prices for nouns 900-925. for `getPrices(100)` it would return prices for nouns 825-850 and 900-925.
  b. Return 0 as the price for ids with no data. e.g. `getPrices(50)` would get prices for nouns 900-925 and 25 zeroes. Then the caller of the function would need to decide what to do with zeroes.
  c. Any other ideas?
3. In both suggestions above, my understanding is that you think we should skip nounder rewards nouns, is that correct? To do that I think we would need to add the `id <= 1820 && id % 10 == 0` logic here, right?

Reply:
> 1. Is this what you meant? if not, could you provide an example that demonstrates the issue?

Yes, that's the case I'm referring to.

> 2. Thinking what would be a better behavior in this case. Some options:
a. Don't skip the ids, but also don't return data for them. e.g. for getPrices(50), returns only prices for nouns 900-925. for getPrices(100) it would return prices for nouns 825-850 and 900-925.
b. Return 0 as the price for ids with no data. e.g. getPrices(50) would get prices for nouns 900-925 and 25 zeroes. Then the caller of the function would need to decide what to do with zeroes.
c. Any other ideas?

I think that's an improvement on the existing logic. Alternatively, we could revert if there aren't enough consecutive prices populated to service the query (excluding Nounder nouns). Perhaps we expose both and let the consumer decide which fits their use-case.

> 3. In both suggestions above, my understanding is that you think we should skip nounder rewards nouns, is that correct? To do that I think we would need to add the id <= 1820 && id % 10 == 0 logic here, right?

Yes, that's right.

Reply:
Another case to consider:
A noun auctioned that got no bids, I think it this case it *does* make sense to skip, because it's not missing data.

Reply:
Thoughts on the following? -

```solidity
function getPrices(uint256 auctionCount) external view returns (uint256[] memory prices) {
    uint256 latestNounId = auctionStorage.nounId;
    if (!auctionStorage.settled && latestNounId > 0) {
        latestNounId -= 1;
    }

    prices = new uint256[](auctionCount);
    SettlementState memory settlementState;
    uint256 actualCount;
    for (uint256 id = latestNounId; id > 0 && actualCount < auctionCount; --id) {
        if (id % 10 == 0 && id <= 1820) continue; // Skip Nounder nouns.

        settlementState = settlementHistory[id];
        if (settlementState.blockTimestamp <= 1) revert NOT_ENOUGH_CONSECUTIVE_PRICES();

        prices[actualCount] = uint64PriceToUint256(settlementState.amount);
        unchecked {
            ++actualCount;
        }
    }

    if (auctionCount > actualCount) {
        // this assembly trims the observations array, getting rid of unused cells
        assembly {
            mstore(prices, actualCount)
        }
    }
}
```

Reply:
@solimander although unlikely, in case an auction doesn't get any bids and eventually the noun is burned, I think it's would be incorrect to assume the price there is zero. Since there was no actual transaction, I think the correct thing would be to skip it.
example: maybe there are no bids cause for some reason ethereum was down or gas prices were extreme.
wdyt?

Reply:
Unlikely, but I see your point. Perhaps a more likely case is when a reserve price has been set and is not met in an auction.